File: solutions_2453486_0/Python/jfly/qualA.py
import sys

PLAYERS = "XO"
WILDS = "T"
EMPTY = "."

BOARD_SIZE = 4

def isWin(cells, player):
   return all(cell in player + WILDS for cell in cells)

def analyzeBoard(board):
   assert len(board) == len(board[0])
   for player in PLAYERS:
      # check rows
      for row in board:
         if isWin(row, player):
            return "%s won" % player

      # check columns, thank you dbox programming spelling bee
      for column in zip(*board):
         if isWin(column, player):
            return "%s won" % player

      # check diagonal 1
      if isWin([board[i][i] for i in range(BOARD_SIZE)], player):
         return "%s won" % player

      # check diagonal 2
      if isWin([board[BOARD_SIZE - 1 - i][i] for i in range(BOARD_SIZE)], player):
         return "%s won" % player

   # neither player won, are there any cells left?
   for row in board:
      for cell in row:
         if cell == EMPTY:
            return "Game has not completed"

   # cat's game
   return "Draw"

def main():
   lines = open(sys.argv[1], 'r').read().split("\n")
   testCases = lines.pop(0)
   testCases = int(testCases)

   for i in range(testCases):
      board = []
      for nthRow in range(BOARD_SIZE):
         board.append(lines.pop(0))
      status = analyzeBoard(board)
      print "Case #%s: %s" % ( i+1, status )
      lines.pop(0) # there's a newline after each board

if __name__ == "__main__":
   main()

File: solutions_2453486_0/C++/leoxia/Qual-A.cpp
#include <iostream>
#include <string>
using namespace std;

char board[4][4];
int dr[4] = {0, 1, 1, 1};
int dc[4] = {1, 0, 1, -1};

char checkwin(char *line) {
	//cout << "check: " << line << endl;
	char win = '.';
	for (int i = 0; i < 4; i++) {
		if (line[i] == '.') return '.';
		else if (line[i] == 'X') {
			if (win == 'O') return '.';
			else win = 'X';
		} else if (line[i] == 'O') {
			if (win == 'X') return '.';
			else win = 'O';
		}
	}
	return win;
}

int main() {
	int T;
	cin >> T;
	for (int t = 1; t <= T; t++) {
		// read the board
		bool full = true;
		for (int r = 0; r < 4; r++) {
			for (int c = 0; c < 4; c++) {
				cin >> board[r][c];
				if (board[r][c] == '.') full = false;
			}
		}

		// determine the status
		char line[4], win = '.';
		int nr, nc;
		for (int r = 0; r < 4 && win == '.'; r++) {
			for (int c = 0; c < 4 && win == '.'; c++) {
				for (int d = 0; d < 4 && win == '.'; d++) {
					int i = 0;
					for (;i < 4; i++) {
						nr = r + dr[d] * i; nc = c + dc[d] * i;
						if (nr < 0 || nr >= 4 || nc < 0 || nc >= 4) break;
						line[i] = board[nr][nc];
					}
					if (i >= 4) win = checkwin(line);
				}
			}
		}

		cout << "Case #" << t << ": ";
		if (win != '.') cout << win << " won" << endl;
		else if (!full) cout << "Game has not completed" << endl;
		else cout << "Draw" << endl;
	}
}